lib/FrontendTool/LoadedModuleTrace.cpp

```diff
@@ -810,3 +813,69 @@ bool swift::emitLoadedModuleTraceIfNeeded(ModuleDecl *mainModule,
   }
   return false;
 }
+
+
+bool swift::emitObjCMessageSendTraceIfNeeded(ModuleDecl *mainModule,
+                                             const FrontendOptions &opts) {
+  ASTContext &ctxt = mainModule->getASTContext();
+  assert(!ctxt.hadError() &&
+         "We should've already exited earlier if there was an error.");
+  class ObjcMethodRefereceCollector: public SourceEntityWalker {
+    std::set<const clang::ObjCMethodDecl*> results;
+    bool visitDeclReference(ValueDecl *D, CharSourceRange Range,
+                            TypeDecl *CtorTyRef, ExtensionDecl *ExtTyRef,
+                            Type T, ReferenceMetaData Data) override {
+      if (!Range.isValid())
+        return true;
+      if (auto *clangD = dyn_cast_or_null<clang::ObjCMethodDecl>(D->getClangDecl())) {
+        results.insert(clangD);
+      }
+      return true;
+    }
+  public:
+    void dump(llvm::raw_ostream &OS) {
+      OS << "[\n";
+      for (const clang::ObjCMethodDecl* clangD: results) {
+        auto &SM = clangD->getASTContext().getSourceManager();
+        clang::SourceLocation Loc = clangD->getLocation();
+        if (!Loc.isValid()) {
+          continue;
+        }
+        OS << "\t{\n";
+        OS << "\t\t\"method_name\": \"" << clangD->getNameAsString() << "\",\n";
```

Comment:
Is just the method name really all that's needed? Is the expectation that the location will unique them?